support pnp resolver

gun-yu opened the pull request:
This PR adds support for PnP resolution.

https://github.com/microsoft/typescript-go/discussions/1875 in the previous discussion, I have internalized pnp-go into the repository.

Please note that the PR has become quite large due to the inclusion of the PnP source and test code. I apologize for the length and appreciate your understanding.

## what is pnp
Yarn Plug’n’Play (PnP) is a dependency resolution system that removes the need for a traditional node_modules folder.
describe at https://github.com/microsoft/typescript-go/issues/460 

## how to support
### add pnp resolution config in host
```go
type host struct {
	orchestrator *Orchestrator
	host         compiler.CompilerHost
	// Caches that last only for build cycle and then cleared out
	extendedConfigCache tsc.ExtendedConfigCache
	sourceFiles         parseCache[ast.SourceFileParseOptions, *ast.SourceFile]
	configTimes         collections.SyncMap[tspath.Path, time.Duration]

	// caches that stay as long as they are needed
	resolvedReferences parseCache[tspath.Path, *tsoptions.ParsedCommandLine]
	mTimes             *collections.SyncMap[tspath.Path, time.Time]
	resolvedReferences  parseCache[tspath.Path, *tsoptions.ParsedCommandLine]
	mTimes              *collections.SyncMap[tspath.Path, time.Time]
	pnpResolutionConfig *pnp.ResolutionConfig
}
```
### add pnp branch in resolver.go
```go
// resolveTypeReferenceDirective
// skip typeRoots because PnP knows exactly where each @types is located.
	if r.resolver.pnpResolutionConfig != nil {
		if resolvedFromNearestNodeModulesDirectory := r.loadModuleFromNearestNodeModulesDirectory(true /*typesScopeOnly*/); !resolvedFromNearestNodeModulesDirectory.shouldContinueSearching() {
			return r.createResolvedTypeReferenceDirective(resolvedFromNearestNodeModulesDirectory, true /*primary*/)
		}
	} else {
		if len(typeRoots) > 0 {
			if r.tracer != nil {
				r.tracer.write(diagnostics.Resolving_with_primary_search_path_0.Format(strings.Join(typeRoots, ", ")))
			}
```

```go
// loadModuleFromNearestNodeModulesDirectory
if r.resolver.pnpResolutionConfig != nil {
			if result := r.loadModuleFromPNP(priorityExtensions, typesScopeOnly); !result.shouldContinueSearching() {
				return result
			}
		} else {
			if result := r.loadModuleFromNearestNodeModulesDirectoryWorker(priorityExtensions, mode, typesScopeOnly); !result.shouldContinueSearching() {
				return result
			}
		}
```

jakebailey commented on internal/module/pnp/utils/portable_path.go:
```diff
@@ -0,0 +1,47 @@
+}
+
+func fromPortablePath(s string) string {
+	if runtime.GOOS != "windows" {
```
Depending on `runtime.GOOS` is a bad idea; all of our other path handling is platform independent.

gun-yu replied:
[305fd49](https://github.com/microsoft/typescript-go/pull/1876/commits/305fd49550f74402b0dd28cffa7adfb4d090ead6)
I removed this file as normalizePath has been removed.

jakebailey commented on internal/vfs/pnpvfs/pnpvfs.go:
```diff
@@ -0,0 +1,399 @@
+	var pathTail string
+
+	// Do a quick check for a ".zip" in the path at all
+	path = strings.ReplaceAll(path, "\\", "/")
```
`path` is guaranteed to not contain any backslashes.

gun-yu replied:
[f0ee55b](https://github.com/microsoft/typescript-go/pull/1876/commits/f0ee55baae946a3fcbace07443488162481dedca)
remove replaceAll

jakebailey commented on internal/vfs/pnpvfs/pnpvfs.go:
```diff
@@ -0,0 +1,399 @@
+		return false
+	}
+
+	_, ok := zip.files[strings.ToLower(pathTail)]
```
I am suspicious about `ToLower` here, compared to `tspath.Path`.

jakebailey commented on internal/vfs/pnpvfs/pnpvfs.go:
```diff
@@ -0,0 +1,399 @@
+	}
+
+	files := make([]string, 0)
+	dirs := make([]string, 0)
```
`make([]string, 0)` is just an allocate-y version of `var files []string`; don't intentionally make an empty slice unless `nil` vs non-`nil` matters.

jakebailey commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module github.com/microsoft/typescript-go
 
-go 1.25
+go 1.25.1
```
Please undo this.

jakebailey commented on internal/module/pnp/lib.go:
```diff
@@ -0,0 +1,491 @@
+	return "", false
+}
+
+var rePNP = regexp.MustCompile(`(?s)(const[\ \r\n]+RAW_RUNTIME_STATE[\ \r\n]*=[\ \r\n]*|hydrateRuntimeState\(JSON\.parse\()'`)
```
This scares me a bit; because of the regex, but also because this reads JS files? Does the spec really require such a thing?

gun-yu replied:
Ultimately, parsing pnp.cjs is necessary to read the metadata, so this part seems unavoidable.
If we parse it using an AST like esbuild does, it might be a bit more reliable — would that be okay?

jakebailey replied:
That makes more sense to me. Though in general I really thought that PnP had been specified to just emit a plain JSON file...

gun-yu replied:
In my project, where pnp.cjs is around 200,000 lines, the regex implementation performed slightly better.

jakebailey replied:
https://github.com/yarnpkg/pnp-rs was happy with a regex, then I don't mind behind happy with a regex. I don't know if this is even in the hot path. My main concern was just whether or not it was "stable".

gun-yu commented:
@jakebailey Thank you for the thoughtful review! I know feedback like this takes a lot of time, and I truly appreciate it. Reading your comments made me realize how unfamiliar I still am with the TypeScript Go codebase—I’m sorry I didn’t reference the existing code more. Consistent code style is especially important in open source, so I’m happy to adopt all of your suggestions. I’ll start by addressing your comments right away.

gun-yu commented on internal/module/pnp/manifest.go:
```diff
@@ -0,0 +1,231 @@
+}
+
+func (r *RegexDef) compile() (*regexp.Regexp, error) {
+	return regexp.Compile(r.Source)
```
Good point! esbuild uses the JS AST for parsing. Since I was primarily focused on porting pnp-rs, I initially brought the code over as-is, but as you mentioned in another comment, there are definitely some risky areas. I believe the TypeScript Go codebase also includes a JS parser, so I’ll try using that for parsing instead.

gun-yu replied:
Thanks for the feedback!
The original `ignorePatternData` looks like this:

```json
"ignorePatternData": "(^(?:\\\\.yarn\\\\/sdks(?:\\\\/(?!\\\\.{1,2}(?:\\\\/|$))(?:(?:(?!(?:^|\\\\/)\\\\.{1,2}(?:\\\\/|$)).)*?)|$))$)",
```

[Esbuild](https://github.com/evanw/esbuild/blob/main/internal/resolver/yarnpnp.go#L327) also compiles and uses the regex in the same way. Since this is part of the Yarn PnP spec, it seems unavoidable to use a regular expression here.

For now, I’ve optimized this commit to avoid the inefficiency of compiling the regex on every use. [f0ee55b](https://github.com/microsoft/typescript-go/pull/1876/commits/f0ee55baae946a3fcbace07443488162481dedca)

gun-yu commented on internal/module/pnp/utils/normalize_path.go:
```diff
@@ -0,0 +1,64 @@
+	"strings"
+)
+
+func NormalizePath(original string) string {
```
[305fd49](https://github.com/microsoft/typescript-go/pull/1876/commits/305fd49550f74402b0dd28cffa7adfb4d090ead6)
I replaced normalizePath with tspath.normalizePath.

gun-yu commented on internal/module/pnp/utils/portable_path.go:
```diff
@@ -0,0 +1,47 @@
+	reUNCWindowsPath  = regexp.MustCompile(`^[\/\\][\/\\](\.[\/\\])?(.*)$`)
+	rePortablePath    = regexp.MustCompile(`^\/([a-zA-Z]:.*)$`)
+	reUNCPortablePath = regexp.MustCompile(`^\/unc\/(\.dot\/)?(.*)$`)
+)
```
[305fd49](https://github.com/microsoft/typescript-go/pull/1876/commits/305fd49550f74402b0dd28cffa7adfb4d090ead6)
I removed this file as normalizePath has been removed.

gun-yu commented on internal/module/pnp/manifest.go:
```diff
@@ -0,0 +1,231 @@
+	"github.com/microsoft/typescript-go/internal/module/pnp/utils"
+)
+
+type Trie[T any] = utils.Trie[T]
```
[305fd49](https://github.com/microsoft/typescript-go/pull/1876/commits/305fd49550f74402b0dd28cffa7adfb4d090ead6)
I removed the utils folder and moved its contents into the pnp package.
Does this direction look okay to you, or would it be better to place it under internal/collections instead?
I’ve implemented only the Trie features I needed for now, so I included it in pnp for the time being.

jakebailey replied:
It could go there, yes.

gun-yu commented on internal/vfs/pnpvfs/pnpvfs.go:
```diff
@@ -0,0 +1,399 @@
+
+func (fs *PNPFS) Chtimes(path string, aTime time.Time, mTime time.Time) error {
+	return fs.inner.Chtimes(path, aTime, mTime)
+}
```
[ae272f1](https://github.com/microsoft/typescript-go/pull/1876/commits/ae272f16e2024b1221878883eafd07dbbef28182)
add assertion on writeFile, Remove, Chtimes

gun-yu commented:
@jakebailey 
I believe I’ve incorporated all the review feedback.
Please let me know if there’s anything I might have overlooked.

jakebailey commented on internal/module/pnp/lib.go:
```diff
@@ -0,0 +1,491 @@
+	Err          error
+}
+
+func (e *FailedManifestHydrationError) Error() string {
```
The trace is for debugging, it's generally okay for it to be long, though there is a balance. How long are you talking?

jakebailey commented on internal/module/pnp/manifest.go:
```diff
@@ -0,0 +1,248 @@
+	IsAlias   bool
+}
+
+func (p *PackageDependency) UnmarshalJSON(data []byte) error {
```
Not strictly required, but all of these UnmarshalJSON functions would perform better and probably be simpler to write using the v2 versions of this, as in `UnmarshalJSONFrom(dec *jsontext.Decoder) error`).

gun-yu replied:
[e697dce](https://github.com/microsoft/typescript-go/pull/1876/commits/e697dcee8beb826f317b153f5f7ea78f61548daa)
I did some work on it, but the code has gotten quite complicated, so I'm not sure if it's heading in the right direction. There might be some minor memory or performance gains, but I haven’t benchmarked it yet.

gun-yu replied:
I did a simple timing test, and the new implementation was about twice as fast. Nice!

jakebailey commented on internal/module/resolver.go:
```diff
@@ -1973,7 +2003,7 @@ func extensionIsOk(extensions extensions, extension string) bool {
 
 func ResolveConfig(moduleName string, containingFile string, host ResolutionHost) *ResolvedModule {
-	resolver := NewResolver(host, &core.CompilerOptions{ModuleResolution: core.ModuleResolutionKindNodeNext}, "", "")
+	resolver := NewResolver(host, &core.CompilerOptions{ModuleResolution: core.ModuleResolutionKindNodeNext}, "", "", nil)
```
This is an interesting one; `nil` here means that one can't load tsconfigs through PnP, which is probably fine, but I am curious how the old PnP patch did this (if at all).

gun-yu replied:
[01b7670](https://github.com/microsoft/typescript-go/pull/1876/commits/01b76704a7280dea56675eaa85f95a0c986795aa)
I’ve modified the interface to inject pnpResolutionConfig as shown in this commit. What do you think about this approach?

jakebailey commented on internal/module/resolver_pnp.go:
```diff
@@ -0,0 +1,52 @@
+
+	// trim trailing slash makes a bug in packageJsonInfoCache.Set
+	// like @emotion/react/ -> @emotion/react after packageJsonInfoCache.Set
+	// check why it's happening in packageJsonInfoCache and need to fix it
```
Yes, because the cache I think assumes normalized paths with trailing trimmed. Definitely need to ensure paths are minimal like that. I think.

gun-yu replied:
[6747e23](https://github.com/microsoft/typescript-go/pull/1876/commits/6747e2343a417e93ed6ca025249ad8f41ebf5a5d)
It’s a bit unfortunate that I have to explain this behavior in a comment, but I’ve written it this way for now.

jakebailey commented on internal/project/ata/ata.go:
```diff
@@ -186,7 +186,7 @@ func (ti *TypingsInstaller) installTypings(
 		logger.Log(fmt.Sprintf("ATA:: Installed typings %v", packageNames))
 		var installedTypingFiles []string
-		resolver := module.NewResolver(ti.host, &core.CompilerOptions{ModuleResolution: core.ModuleResolutionKindNodeNext}, "", "")
+		resolver := module.NewResolver(ti.host, &core.CompilerOptions{ModuleResolution: core.ModuleResolutionKindNodeNext}, "", "", nil)
```
Unless you plan to do it in this PR, we this should have a comment in the code.

gun-yu replied:
[6b2f0c1](https://github.com/microsoft/typescript-go/pull/1876/commits/6b2f0c186e17a6feacc6c7860670d3418cfa6111)
I’ve added a comment noting that automatic type installation is not supported in the PnP environment.

gun-yu commented on internal/compiler/pnp.go:
```diff
@@ -0,0 +1,54 @@
+import (
+	"io/fs"
+	"os"
+	"path/filepath"
```
[1041038](https://github.com/microsoft/typescript-go/pull/1876/commits/10410382498dc30d98b70c2b29673a50fe8aae93)
I replaced the implementation to use Yarn’s FindManifest API and switched from using the OS-based implementation to a VFS-based one.

gun-yu commented:
> > I noticed that there’s another PR with a similar goal: #1966. To avoid overlapping work, could you please advise on the best way to move forward?
> 
> Hey @gun-yu! We were aware of your PR, and we are totally open to collaborate on this feature, especially since we both implemented features that the other didn't focus on. We already sent you a couple of email on this 🙇

Thank you for your response! I’d be very happy to collaborate if possible.

I also took a look at your PR — I saw that it already implements both the PnP VFS and metadata parsing.
I was particularly thinking about where the PnP VFS should fit in the project structure, and I found the approach in that PR quite inspiring.

I just noticed your email recently — I don’t check my inbox very often, so I apologize for the delayed reply!

jakebailey commented:
Of course, sure
